src/lib.rs

```diff
@@ -686,10 +920,269 @@ struct ResUnit<R: gimli::Reader> {
+where
+    L: LookupContinuation,
+    F: FnOnce(L::Output) -> T,
+{
+    original: L,
+    mutator: F,
+}
+
+impl<T, L, F> LookupContinuation for MappedLookup<T, L, F>
+where
+    L: LookupContinuation,
+    F: FnOnce(L::Output) -> T,
+{
+    type Output = T;
+    type Buf = L::Buf;
+
+    fn resume(self, v: Option<Arc<gimli::Dwarf<Self::Buf>>>) -> LookupResult<Self> {
+        match self.original.resume(v) {
+            ControlFlow::Break(t) => ControlFlow::Break((self.mutator)(t)),
+            ControlFlow::Continue((load, continuation)) => ControlFlow::Continue((
+                load,
+                MappedLookup {
+                    original: continuation,
+                    mutator: self.mutator,
+                },
+            )),
+        }
+    }
+}
+
+struct LoopingLookup<T, L, F>
+where
+    L: LookupContinuation,
+    F: FnMut(L::Output) -> ControlFlow<T, LookupResult<L>>,
+{
+    continuation: L,
+    mutator: F,
+}
+
+impl<T, L, F> LoopingLookup<T, L, F>
+where
+    L: LookupContinuation,
+    F: FnMut(L::Output) -> ControlFlow<T, LookupResult<L>>,
+{
+    fn new_complete(t: T) -> LookupResult<Self> {
+        ControlFlow::Break(t)
+    }
+
+    fn new_lookup(mut r: LookupResult<L>, mut mutator: F) -> LookupResult<Self> {
+        // Drive the loop eagerly so that we only ever have to represent one state
+        // (the r == ControlFlow::Continue state) in LoopingLookup.
+        let (load, continuation) = loop {
+            match r {
+                ControlFlow::Break(l) => match mutator(l) {
+                    ControlFlow::Break(t) => return ControlFlow::Break(t),
+                    ControlFlow::Continue(r2) => {
+                        r = r2;
+                    }
+                },
+                ControlFlow::Continue((load, continuation)) => break (load, continuation),
+            }
+        };
+
+        ControlFlow::Continue((
+            load,
+            LoopingLookup {
+                continuation,
+                mutator,
+            },
+        ))
+    }
+}
+
+impl<T, L, F> LookupContinuation for LoopingLookup<T, L, F>
+where
+    L: LookupContinuation,
+    F: FnMut(L::Output) -> ControlFlow<T, LookupResult<L>>,
+{
+    type Output = T;
+    type Buf = L::Buf;
+
+    fn resume(self, v: Option<Arc<gimli::Dwarf<Self::Buf>>>) -> LookupResult<Self> {
+        let r = self.continuation.resume(v);
+        LoopingLookup::new_lookup(r, self.mutator)
+    }
 }
 
 impl<R: gimli::Reader> ResUnit<R> {
+    fn dwarf_and_unit_dwo<'unit, 'ctx: 'unit>(
+        &'unit self,
+        ctx: &'ctx Context<R>,
+    ) -> LookupResult<
+        SimpleLookup<
+            Result<(&'unit RawDwarf<R>, &'unit gimli::Unit<R>), Error>,
+            R,
+            impl FnOnce(
+                Option<Arc<gimli::Dwarf<R>>>,
+            ) -> Result<(&'unit RawDwarf<R>, &'unit gimli::Unit<R>), Error>,
+        >,
+    > {
+        loop {
+            break SimpleLookup::new_complete(match self.dwo.borrow() {
```

Comment:
This `new_complete` is a bit far removed from the code that uses it. Either move it inside the match, or move the last 3 cases to here.